common/oauth.go

```diff
@@ -99,11 +110,62 @@ func newOAuthClient(ctx context.Context, params *oauthClientParams) Authenticate
 	// This is how the key refresher accepts a custom http client
 	ctx = context.WithValue(ctx, oauth2.HTTPClient, params.client)
 
+	tokenSource := getTokenSource(ctx, params)
+	if params.updateToken != nil {
+		tokenSource = &writeBackTokenSource{
+			update:      params.updateToken,
+			lastKnown:   params.token,
+			tokenSource: tokenSource,
+		}
+	}
+
 	// Returns a new client which automatically refreshes the access token
 	// whenever the current one expires.
+	return oauth2.NewClient(ctx, tokenSource)
+}
+
+func getTokenSource(ctx context.Context, params *oauthClientParams) oauth2.TokenSource { //nolint:ireturn
 	if params.tokenSource != nil {
-		return oauth2.NewClient(ctx, params.tokenSource)
-	} else {
-		return oauth2.NewClient(ctx, params.config.TokenSource(ctx, params.token))
+		return params.tokenSource
+	}
+
+	return params.config.TokenSource(ctx, params.token)
+}
+
+type writeBackTokenSource struct {
```

Comment:
Let's make this struct name more generic? Since persisting token update is just one use case, there might be other use cases of a callback function.

Reply:
How about `observableTokenSource`?